### PR TITLE
Fixes #9427 - Return meaningful errors from subnets/freeip 

### PR DESCRIPTION
**Pull request by ShimShtein**
and parse the error response to the UI

### CODE REVIEW

File: app/controllers/subnets_controller.rb [hunk 1]
```diff
@@ -40,21 +40,20 @@ def destroy
...
       else
-        # we don't want any failures if we failed to query our proxy
-        head :status => 200
+        not_found
       end
     end
   rescue => e
     logger.warn "Failed to query subnet #{s} for free ip: #{e}"
-    head :status => 500
+    generic_exception e
```
Comment (domcleal) [inline app/controllers/subnets_controller.rb]: generic_exception doesn't need to be called explicitly, it should be rescued (see app controller).  But there's also process_ajax_error in app controller, which is used in hosts controller, can that be reused?

**Comment by dLobatog** [comment]
@ShimShtein cool, could you add some tests for the new behavior?

**Comment by ShimShtein** [comment]
Thanks @eLobato, added tests for the feature.

**Comment by ares** [comment]
[test]

**Comment by ares** [comment]
rubocops issue seems unrelated

File: app/controllers/subnets_controller.rb [hunk 2]
```diff
@@ -40,21 +40,23 @@ def destroy
...
   # query our subnet dhcp proxy for an unused IP
   def freeip
-    not_found and return unless (s=params[:subnet_id].to_i) > 0
+    invalid_request and return unless (s=params[:subnet_id].to_i) > 0
     organization = params[:organization_id].blank? ? nil : Organization.find(params[:organization_id])
     location = params[:location_id].blank? ? nil : Location.find(params[:location_id])
     Taxonomy.as_taxonomy organization, location do
-      not_found and return unless (subnet = Subnet.authorized(:view_subnets).find(s))
+      unless (subnet = Subnet.authorized(:view_subnets).find(s))
+        render_403 #returns false, cannot put it in a oneliner
```
Comment (dLobatog) [inline app/controllers/subnets_controller.rb]: Why 403? `not_found` is what Foreman is expected to return when it cannot find a resource (in this case Subnets). I think it's not possible to even fail the `.authorized` check here through the UI, as in order to call freeip you first have to select a subnet.

**Comment by dLobatog** [comment]
@ShimShtein Thanks, other than the comment above I tested it and it works pretty well :+1: .

File: app/controllers/subnets_controller.rb (hunk 2 above)
Comment (ShimShtein) [inline app/controllers/subnets_controller.rb]: not_found means that the resource is not there, it's not related to the context of the user.
403 means forbidden - which means that under the current context the action is not possible, but if the context will be different, the action will be available.
I think the proper use of not_found will be if the system just can't find available IP, just as I did [three lines below](https://github.com/ShimShtein/foreman/blob/9427/app/controllers/subnets_controller.rb#L54)

**Comment by ares** [comment]
[test]

**Comment by dLobatog** [comment]
Merged as b9521a8dc7f4e61a011cabbfdfe78657bd3c24d2, thanks @ShimShtein

**Pull request closed without merging**